File: Session_06_DynamicProgramming_Classic/multi_language_implementations/longest_common_subsequence.js
const readline = require('readline');

function solveLCS(s1, s2) {
    const m = s1.length;
    const n = s2.length;
    const dp = Array.from({ length: m + 1 }, () => new Array(n + 1).fill(0));

    for (let i = 0; i <= m; i++) {
        for (let j = 0; j <= n; j++) {
            if (i === 0 || j === 0) {
                dp[i][j] = 0;
            } else if (s1[i - 1] === s2[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1] + 1;
            } else {
                dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
    }
    console.log(`LCS Length: ${dp[m][n]}`);

    let index = dp[m][n];
    const lcsStr = new Array(index);
    let i = m, j = n;
    while (i > 0 && j > 0) {
        if (s1[i - 1] === s2[j - 1]) {
            lcsStr[index - 1] = s1[i - 1];
            i--;
            j--;
            index--;
        } else if (dp[i - 1][j] > dp[i][j - 1]) {
            i--;
        } else {
            j--;
        }
    }
    console.log(`LCS String: ${lcsStr.join("")}`);
}

function main() {
    const s1 = "STONE";
    const s2 = "LONEST";
    solveLCS(s1, s2);
    console.log();

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    rl.question("Press Enter to exit...", () => {
        rl.close();
    });
}

main();

File: Session_06_DynamicProgramming_Classic/multi_language_implementations/longest_common_subsequence.py
import sys

def solve_lcs(s1, s2):
    m = len(s1)
    n = len(s2)
    dp = [[0] * (n + 1) for _ in range(m + 1)]
    for i in range(m + 1):
        for j in range(n + 1):
            if i == 0 or j == 0:
                dp[i][j] = 0
            elif s1[i - 1] == s2[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
                
    print(f"LCS Length: {dp[m][n]}")
    
    index = dp[m][n]
    lcs_chars = [""] * index
    i, j = m, n
    while i > 0 and j > 0:
        if s1[i - 1] == s2[j - 1]:
            lcs_chars[index - 1] = s1[i - 1]
            i -= 1
            j -= 1
            index -= 1
        elif dp[i - 1][j] > dp[i][j - 1]:
            i -= 1
        else:
            j -= 1
            
    lcs_str = "".join(lcs_chars)
    print(f"LCS String: {lcs_str}")

def main():
    s1 = "STONE"
    s2 = "LONEST"
    solve_lcs(s1, s2)
    print()
    input("Press Enter to exit...")

if __name__ == "__main__":
    main()
